ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -1247,133 +1247,109 @@ int ompi_coll_base_allreduce_intra_redscat_allgather(
-    tmp_buf_raw = (char *) malloc(dsize);
-    if (NULL == tmp_buf_raw) {
-        return OMPI_ERR_OUT_OF_RESOURCE;
-    }
-
-    tmp_buf = tmp_buf_raw - gap;
-
-    /* Requests for send to AND receive from everyone else */
-    int reqs_needed = (comm_size - 1) * 2;
-    reqs = ompi_coll_base_comm_get_reqs(module->base_data, reqs_needed);
-
-    const ptrdiff_t incr = extent * count;
-
-    /* Exchange data with peer processes, excluding self */
-    int req_index = 0, peer_rank = 0;
-    for (int i = 1; i < comm_size; ++i) {
-        /* Start at the next rank */
-        peer_rank = (rank + i) % comm_size;
-
-        /* Prepare for the next receive buffer */
-        if (0 == peer_rank && rbuf != send_buf) {
-            /* Optimization for Rank 0 - its data will always be placed at the beginning of local
-             * reduce output buffer.
-             */
-            tmp_recv = rbuf;
-        } else {
-            tmp_recv = tmp_buf + (peer_rank * incr);
-        }
 
-        err = MCA_PML_CALL(irecv(tmp_recv, count, dtype, peer_rank, MCA_COLL_BASE_TAG_ALLREDUCE,
-                                 comm, &reqs[req_index++]));
-        if (MPI_SUCCESS != err) {
-            goto err_hndl;
-        }
+    int rank = ompi_comm_rank(comm);
+    int size = ompi_comm_size(comm);
 
-        err = MCA_PML_CALL(isend(send_buf, count, dtype, peer_rank, MCA_COLL_BASE_TAG_ALLREDUCE,
-                                 MCA_PML_BASE_SEND_STANDARD, comm, &reqs[req_index++]));
-        if (MPI_SUCCESS != err) {
-            goto err_hndl;
-        }
+    sendtmpbuf = (char*) sbuf;
+    if( sbuf == MPI_IN_PLACE ) {
+        sendtmpbuf = (char *)rbuf;
+    }
+    ptrdiff_t buf_size, gap = 0;
+    buf_size = opal_datatype_span(&dtype->super, (int64_t)count * size, &gap);
+    partial_buf = (char *) malloc(buf_size);
+    partial_buf_start = partial_buf - gap;
+    buf_size = opal_datatype_span(&dtype->super, (int64_t)count, &gap);
+    buffer1 = (char *) malloc(buf_size);
+    buffer1_start = buffer1 - gap;
+
+    err = ompi_datatype_copy_content_same_ddt(dtype, count,
+                                              (char*)buffer1_start,
+                                              (char*)sendtmpbuf);
+    if (MPI_SUCCESS != err) { line = __LINE__; goto err_hndl; }
+
+    // apply allgather data so that each rank has a full copy to do reduce (trade bandwidth for better latency)
+    err = comm->c_coll->coll_allgather(buffer1_start, count, dtype,
+                                       partial_buf_start, count, dtype,
+                                       comm, comm->c_coll->coll_allgather_module);
+    if (MPI_SUCCESS != err) { line = __LINE__; goto err_hndl; }
+
+    for(int target = 1; target < size; target++)
+    {
+        ompi_op_reduce(op,
+                       partial_buf_start + (ptrdiff_t)target * count * extent,
+                       partial_buf_start,
+                       count,
+                       dtype);
     }
```

Comment:
nit Code formatting. Opening brace should be on the same line as `for`. And there should be a space between `for` and `(`